grand_challenge_forge/partials/example-evaluation-method/example-evaluation-method{{cookiecutter._}}/helpers.py

```diff
@@ -137,28 +118,47 @@ def sigchld_handler(*_, **__):
                     result = future.result()
                     results.append(result)
                 except Exception as e:
-                    handle_error(e, prediction=future_to_predictions[future])
+                    errors.append((future_to_predictions[future], e))
+
+                    if not caught_exception:  # Hard stop
+                        caught_exception = True
+
+                        executor.shutdown(wait=False, cancel_futures=True)
+                        _terminate_child_processes()
         finally:
-            terminating_child_processes = True
+            # Be aggresive in cleaning up any left-over processes
             _terminate_child_processes()
 
 
 def _terminate_child_processes():
-    current_process = psutil.Process(os.getpid())
-    children = current_process.children(recursive=True)
+    process = psutil.Process(os.getpid())
+    children = process.children(recursive=True)
     for child in children:
         try:
             child.terminate()
         except psutil.NoSuchProcess:
             pass  # Not a problem
 
     # Wait for processes to terminate
-    gone, still_alive = psutil.wait_procs(children, timeout=5)
+    _, still_alive = psutil.wait_procs(children, timeout=5)
 
     # Forcefully kill any remaining processes
     for p in still_alive:
-        print(f"Forcefully killing child process {p.pid}")
         try:
             p.kill()
         except psutil.NoSuchProcess:
             pass  # That is fine
+
+    # Finally, prevent zombies by waiting for all child processes
+    try:
+        os.waitpid(-1, 0)
+    except ChildProcessError:
+        pass  # No child processes, that if fine
+
+
+def listen_to_children_errors():
+    def handler(*_, **__):
+        print("A child failed, terminating all other children")
+        _terminate_child_processes()
+
+    signal.signal(signal.SIGCHLD, handler)
```

Comment:
Is this used anywhere? I cannot find a place where this is called/registered? It seems to have been replaced by line 123, is that correct?

Reply:
Thanks for reviewing!

This function is used on the evaluate.py.j2 line 46. The addition is mainly meant to automagically catch errors of grandchildren of the main evaluation thread (i.e. children of the execution pool workers).